ql/src/java/org/apache/hadoop/hive/ql/plan/mapper/PlanMapper.java

```diff
@@ -200,54 +199,30 @@ public void merge(Object o1, Object o2) {
   }
 
   private void link(Object o1, Object o2, boolean mayMerge) {
-
-    Set<Object> keySet = Collections.newSetFromMap(new IdentityHashMap<Object, Boolean>());
-    keySet.add(o1);
-    keySet.add(o2);
-    keySet.add(getKeyFor(o1));
-    keySet.add(getKeyFor(o2));
-
-    Set<EquivGroup> mGroups = Collections.newSetFromMap(new IdentityHashMap<EquivGroup, Boolean>());
-
-    for (Object object : keySet) {
-      EquivGroup group = objectMap.get(object);
-      if (group != null) {
-        mGroups.add(group);
-      }
-    }
-    if (mGroups.size() > 1) {
+    final EquivGroup group1 = objectMap.get(o1);
+    final EquivGroup group2 = objectMap.get(o2);
+    if (group1 != null && group2 != null && group1 != group2) {
       if (!mayMerge) {
         throw new RuntimeException("equivalence mapping violation");
```

Comment:
If we touch this change please the exception message to Uppercase the first word